### PR TITLE
Rs256 jwt signature verification

### LINKED ISSUE #421: Add built-in function(s) to verify JWT signatures
The current `io.jwt.decode` built-in doesn't perform any crypto verification of the signature. As mentioned in #417, it would be nice if OPA could verify JWT signatures.

I could see this working one of a few ways (in order of preference):

1. Add new crypto built-ins to verify signatures independent of JWTs.
2. Modify the existing JWT decode built-in to to support verification (e.g., add a `key` param that if set would enable verification).
3. Add a new JWT decode built-in that also verifies the signature.

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
A builtin to verify RS256 JWT signatures
Issue - https://github.com/open-policy-agent/opa/issues/421
Fixes #421

### CODE REVIEW

File: topdown/tokens_verify.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+	block, rest := pem.Decode([]byte(certificate))
+
+	if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
+		return nil, fmt.Errorf("failed to decode PEM block containing certificate")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse PEM certificate %v", err)
```
Comment (tsandall) [inline topdown/tokens_verify.go]: See `errors.Wrapf` for returning wrapped errors. This way the embedded error formatting is more consistent. For example, I would use `errors.Wrapf(err, "PEM parse error")`.

File: topdown/tokens_verify.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	case 2:
+		data += "==" // 2 pad chars
+	case 3:
+		data += "=" // 1 pad char
+	}
+
+	return base64.StdEncoding.DecodeString(data)
+}
+
+// encode encodes given byte array to base64url string
```
Comment (tsandall) [inline topdown/tokens_verify.go]: Can this be replaced with `base64.URLEncoding.EncodeToString`?

File: topdown/tokens_verify.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// getInputSha returns the SHA256 checksum of the input
+func getInputSHA(input []byte) (hash []byte) {
+	hasher := sha256.New()
+	hasher.Write(input)
+	return hasher.Sum(nil)
+}
+
+// decode decodes base64url string to byte array
```
Comment (tsandall) [inline topdown/tokens_verify.go]: Can this be replaced with base64.URLEncoding.Decode?

File: topdown/tokens_verify.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+func builtinJWTVerifyRS256(a ast.Value, b ast.Value) (ast.Value, error) {
+
+	// Process the token
+	astToken, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+	token := string(astToken)
+	if !strings.Contains(token, ".") {
+		return nil, errors.New("encoded JWT has no period separators")
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply topdown/tokens_verify.go]: Removed

File: topdown/tokens_verify.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	// Get public key
+	publicKey := cert.PublicKey.(*rsa.PublicKey)
+
+	// Validate the JWT signature
+	err = rsa.VerifyPKCS1v15(publicKey, crypto.SHA256, getInputSHA(headerPayload), []byte(signature))
+
+	if err != nil {
+		return ast.String("RS256 JWT signature verification failed"), nil
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply topdown/tokens_verify.go]: Returning an error now

File: topdown/tokens_verify.go (hunk 1 above)
Comment (ashutosh-narkar) [reply topdown/tokens_verify.go]: Used errors.Wrap

File: topdown/tokens_verify.go (hunk 2 above)
Comment (ashutosh-narkar) [reply topdown/tokens_verify.go]: Yes it can. Do we want to ?

File: topdown/tokens_verify.go (hunk 3 above)
Comment (ashutosh-narkar) [reply topdown/tokens_verify.go]: I need conversion from string to []byte . So "Decode" won't help.

File: topdown/tokens_verify.go (hunk 2 above)
Comment (tsandall) [reply topdown/tokens_verify.go]: Yes, let's use the standard library implementation instead.

File: topdown/tokens_verify.go (hunk 3 above)
Comment (tsandall) [reply topdown/tokens_verify.go]: I think `base64.URLEncoding.DecodeString` will work.

**Pull request merged**